src/System.IO/src/System/IO/TextWriter.cs

```diff
@@ -391,11 +391,43 @@ public virtual void WriteLine(decimal value)
         //
         public virtual void WriteLine(string value)
         {
-            if (value != null)
+            if (value==null)
             {
-                Write(value);
+                WriteLine();
             }
-            Write(CoreNewLineStr);
+            else
+            {
+                // We'd ideally like WriteLine to be atomic, in that one call
```

Comment:
Seems like the base TextWriter should just do the simple thing:
```C#
if (value != null) Write(value);
Write(CoreNewLineStr);
```
and then derived classes can override WriteLine if it's relevant.  As it currently stands, since StreamWriter and StringWriter both buffer (and don't override WriteLine), it actually seems more expensive to do this more complicated implementation in the base class.  I'm not sure if that's what you both are suggesting in your most recent comments, but if so, +1 :smile: